### PR TITLE
Make pushing back to vectors more efficient

### PR DESCRIPTION
**Pull request by AtariDreams**
In these places, emplace_back seems to be more efficient than push_back, and the better choice.

We can also improve performance further by reserving the size we know the vector will be so we do not have to waste time resizing.

### CODE REVIEW

**Comment by JavierMatosD** [comment]
How did you determine what to refactor? Are you using a tool we could potentially use for our tests?

File: src/vcpkg-test/binarycaching.cpp [hunk 1]
```diff
@@ -458,7 +458,7 @@ Description: a spiffy compression library wrapper
     auto maybe_scf = SourceControlFile::parse_control_file("", std::move(*pghs.get()));
     REQUIRE(maybe_scf.has_value());
     SourceControlFileAndLocation scfl{std::move(*maybe_scf.get()), Path()};
-    plan.install_actions.push_back(InstallPlanAction());
+    plan.install_actions.emplace_back();
```
Comment (BillyONeal) [inline src/vcpkg-test/binarycaching.cpp:461]: I double checked that `InstallPlanAction` is not an aggregate ✅

File: src/vcpkg/base/files.cpp [hunk 2]
```diff
@@ -2358,7 +2358,7 @@ namespace vcpkg
         virtual std::vector<Path> get_files_recursive(const Path& dir, std::error_code& ec) const override
         {
             std::vector<Path> result;
-            Path out_base = dir;
+            const Path& out_base = dir;
             get_files_recursive_impl(result, dir, out_base, ec, true, true, true);
```
Comment (BillyONeal) [inline src/vcpkg/base/files.cpp:2362]: This seems wrong: Either the caller explicitly needed this to be a copy to call a non-`const&` overload of `get_files_recursive_impl` or similar, or it should just be:

```suggestion
            get_files_recursive_impl(result, dir, dir, ec, true, true, true);
```

Ditto below.

Comment (BillyONeal) [reply src/vcpkg/base/files.cpp:2362]:
```
        static void get_files_recursive_impl(std::vector<Path>& result,
                                             const Path& base,
                                             const Path& out_base,
                                             std::error_code& ec,
                                             bool want_directories,
                                             bool want_regular_files,
                                             bool want_other)
```

Pretty sure it's the latter...

File: src/vcpkg/build.cpp [hunk 3]
```diff
@@ -624,7 +624,7 @@ namespace vcpkg
                               });
         // Make sure GIT could be found
         const Path& git_exe_path = paths.get_tool_exe(Tools::GIT, stdout_sink);
-        out_vars.push_back({"GIT", git_exe_path});
```
Comment (BillyONeal) [inline src/vcpkg/build.cpp:627]: I confirmed that `CMakeVariable` is not an aggregate. ✅

**Review by BillyONeal — CHANGES_REQUESTED** [review]

**Comment by AtariDreams** [comment]
> Hi @AtariDreams, thank you for the contributions. How did you determine what to refactor? Are you using a tool we could potentially use for our tests?

clang-tidy checks for modernize-use-emplace.

**Comment by AtariDreams** [comment]
Fixed!

**Comment by AtariDreams** [comment]
@BillyONeal fixed!

File: src/vcpkg/base/files.cpp (hunk 2 above)
Comment (BillyONeal) [reply src/vcpkg/base/files.cpp]: Other occurrences aren't fixed.

**Review by BillyONeal — CHANGES_REQUESTED** [review]
Other occurrences aren't fixed

**Comment by AtariDreams** [comment]
@BillyONeal Done!

**Review by BillyONeal — APPROVED** [review]

**Comment by BillyONeal** [comment]
Thanks :)

**Pull request merged**